drivers/dose/dose.c

```diff
@@ -107,11 +111,88 @@ static inline void _disable_sense(dose_t *ctx)
 #endif
 }
 
+#ifdef MODULE_DOSE_WATCHDOG
+static unsigned _watchdog_users;
+static dose_t *_dose_base;
+static uint8_t _dose_numof;
+
+static inline void _watchdog_start(void)
+{
+    if (_watchdog_users) {
+        return;
+    }
+
+    _watchdog_users++;
+    timer_start(DOSE_TIMER_DEV);
+}
+
+static inline void _watchdog_stop(void)
+{
+    if (_watchdog_users == 0 || --_watchdog_users) {
+        return;
+    }
+
+    timer_stop(DOSE_TIMER_DEV);
+}
+
+static void _dose_watchdog_cb(void *arg, int chan)
+{
+    (void) chan;
+    (void) arg;
+
+    for (unsigned i = 0; i < _dose_numof; ++i) {
+        dose_t *ctx = &_dose_base[i];
+
+        switch (ctx->state) {
+        case DOSE_STATE_RECV:
+            if (ctx->recv_buf_ptr_last != ctx->recv_buf_ptr) {
+                ctx->recv_buf_ptr_last = ctx->recv_buf_ptr;
+                break;
+            }
+
+            if (ctx->flags & DOSE_FLAG_RECV_BUF_DIRTY) {
+                break;
+            }
+
+            DEBUG_PUTS("timeout");
+            state(&_dose_base[i], DOSE_SIGNAL_XTIMER);
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+static void _watchdog_init(unsigned timeout_us)
+{
+    timer_init(DOSE_TIMER_DEV, US_PER_SEC, _dose_watchdog_cb, NULL);
+    timer_set_periodic(DOSE_TIMER_DEV, 0, timeout_us, TIM_FLAG_RESET_ON_MATCH);
+    timer_stop(DOSE_TIMER_DEV);
+}
+#else
+static inline void _watchdog_start(void) {}
+static inline void _watchdog_stop(void) {}
+#endif
+
 static dose_signal_t state_transit_blocked(dose_t *ctx, dose_signal_t signal)
 {
+    (void) signal;
     uint32_t backoff;
+
+    backoff = random_uint32_range(xtimer_usec_from_ticks(xtimer_min_timeout),
+                                  2 * ctx->timeout_base);
+    xtimer_set(&ctx->timeout, backoff);
```

Comment:
I see that the backoff scheme has been changed, before there was a bias towards receiving nodes, can you explain this change a bit?

Reply:
Thanks!